src/clojure/nrepl/middleware/session.clj

```diff
@@ -192,8 +192,16 @@
                       (nil? current) :idle
                       (and (or (nil? exec-id) (= current exec-id)) ; cas only checks identity, so check equality first
                            (compare-and-set! running current nil))
-                      (do
-                        (doto ^Thread @thread .interrupt .stop)
+                      (let [t ^Thread @thread]
+                        ;; First interrupt the thread. Then wait for a timeout
+                        ;; if not terminated, do a hard stop
+                        (.interrupt t)
+                        (Thread/sleep 100)
```

Comment:
Probably it's makes sense to duplicate the explanations from your last comment here as well.

Reply:
Spun this out into its own fn, with docs added.

Also updated `close` to use the same logic. Think it makes sense.